### ISSUE TITLE
Tutorial: problem in extracting Masks

### ISSUE THREAD

**Issue by Emmavb1988** [jianhuupenn/MorphLink#2]
Thank you very much for MorphLink and providing the step-by-step tutorial.

I am running into a problem when extracting the masks in step5 and step6.
This is most likely resulting from the refinement of the segmentation, since the patchx_pred_refined.npy online consist of [0], while the patchx_pred.npy has 10 unique values [0 1 2 3 4 5 6 7 8 9]. Setting the refine to FALSE does not resolve it. I am using the toydata.

Thank you very much in advance for your help.

**Comment by jinghuang-stats** [comment]
Hi Emma, 
Thanks so much for your interest in MorphLink!
May I ask whether the issue occurred during **Step 4.3 (segment each patch into masks)** when running the following function?
`mph.step4_Segmentation(plot_dir=plot_dir, n_clusters=10, refine=True, refine_threshold=4)`
Could you please share any intermediate results or outputs you obtained when running the toy dataset? That would help me better understand and troubleshoot the problem.
As an additional check, please make sure that your Python environment and dependencies are properly set up to avoid some unexpected errors.

**Comment by Emmavb1988** [comment]
Sorry for not clarifying this better. I have attached a pdf of the Jupyter notebook with output. I have used the code in 1.3 to create my environment and in which I installed all dependencies. 

In addition;
I actually do not receive an error until I try to plot some sample masks with (cell [30])
```
# Plot some sample masks for visuallization
num_samples = 3 # the number of samples for each mask
for channel in range(masks.shape[0]):
    ret_img=mph.mask_example(channel, img_adata, patch_info, patches, masks, plot_dir=plot_dir+"/figures", num_samples=num_samples, filter_mask_area=True)
    ret_img_cvt=cv2.cvtColor(ret_img, cv2.COLOR_BGR2RGB)
    plt.imshow(ret_img_cvt)
    plt.axis('off')
    plt.show()
    plt.close()
```
However, I think this results from the fact that in step 4.3 at 
`mph.step4_Segmentation(plot_dir=plot_dir, n_clusters=10, refine=True, refine_threshold=4)`
the patchx_pred_refined.npy are generated, but only consist of [0], resulting in only 1 mask

[MorphLink_tutorial.html](https://github.com/user-attachments/files/23121538/MorphLink_tutorial.html)

I hope this clarifies it. Thank you again for your help.

**Comment by jinghuang-stats** [comment]
Hi Emma,

Thank you for your clarifications!
In your generated patchx_pred_refined.npy, could you please check how many clusters were generated for the patch? You can run the following commands to confirm:
```
patch_test=np.load("./seg_results/patch0_pred_refined.npy")
np.unique(patch_test)
```
The output should contain 10 clusters, corresponding to the parameter `n_clusters=10` (its default value). 
If you find only one cluster was generated, it likely means that the **KMeans** function did not run successfully:
`kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(pixel_values)`  
This could happen if `KMeans` was not properly imported:
`from sklearn.cluster import KMeans`
In that case, it's an environment setup issue. I'll add `KMeans` to the required package dependencies to prevent this issue in the future. 

If your output does include 10 clusters, could you please also check whether your results folder contains the following files:

- `masks_30_60.npy`
- `masks_index_30_60.pkl`
These files confirm that `mph.step5_Extract_Masks` ran successfully. You should then be able to visualize the preliminary segmented masks using `mph.step6_Plot_Masks`, which will generate a folder named "**masks_30_60**" containing the segmented mask outputs (as shown in the attached figures).

<img width="329" height="219" alt="Image" src="https://github.com/user-attachments/assets/d5165337-08b8-4b8d-8264-41219fee84b8" />

**Comment by Emmavb1988** [comment]
Thank you for your reply.
I have double checked if KMeans was properly imported and did some test to see if it runs properly, and it does.

However, when checking the number of clusters I get this output:

<img width="1264" height="116" alt="Image" src="https://github.com/user-attachments/assets/f7e4b2de-889d-4964-9f2b-28c19e6aa391" />

Indicating that I don't have 10 clusters. 
I have the `masks_30_60.npy` and the `masks_index_30_60.pkl` files in my results folder.
However, I only have one` mask0.png` in "**masks_30_60**" folder.

**Comment by jinghuang-stats** [comment]
Hi Emma,
Thanks so much for checking. Now we can locate the issue to be from the resulting number of clusters. It's indeed unusual that `KMeans` runs without error but doesn't produce the expected number of clusters.
To help pinpoint the cause, I've uploaded the intermediate results (including **patches** and **masks**) to the [shared folder](https://drive.google.com/drive/folders/1fzeyG0rbpl83AL0Qu15Pu0yDg48cJXxL?usp=sharing). You can also find this link in the GitHub tutorial folder under **results_download.md**.
Since you're starting from the same toy dataset, your segmented patches should look very similar to the ones I shared. Please try the following commands using the downloaded `patches.npy` to verify whether your `KMeans` function produces 10 clusters as expected:
```
patches=np.load(plot_dir+"/results/patches.npy")
i=0 
patch=patches[i]
pixel_values = patch.reshape((-1, 3))
pixel_values = np.float32(pixel_values)
n_clusters=10 
kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(pixel_values)
pred=kmeans.labels_
print(np.unique(pred))
```
You can adjust the patch index `i` and the number of clusters `n_clusters` to test whether your `KMeans` function runs properly. I have attached my testing results for your reference. 

- If the output shows the correct number of clusters using my shared patches, the issue is likely at the very beginning of segmenting patches **(Section 4.2: Patch Split)**.
- In that case, please double-check whether the following function ran successfully:
`patches=mph.patch_split_for_ST(img=img, patch_size=patch_size, spot_info=gene_adata.obs, x_name="pixel_x", y_name="pixel_y")`
Please give this a quick check when you can -- it'll help narrow down whether the issue lies in `KMeans` itself or earlier segmenting. 
<img width="508" height="122" alt="Image" src="https://github.com/user-attachments/assets/26e1cb12-797c-4dc4-b15e-7a6309dd49f0" />

**Comment by Emmavb1988** [comment]
Thank you again for your quick reply and options to test.

I used your `patches.npy` as suggested and ran your code provided in your comment above and I do get the correct number of clusters depending on the` n_clusters` I set.

I have checked the patches I generated in 4.2 and see below, that this also seems to produce a workable patches.npy:

<img width="1250" height="821" alt="Image" src="https://github.com/user-attachments/assets/58a1f4d6-a995-437c-ba5e-69cfa455c746" />

Indeed, when running the code below to check the KMeans clustering on my own generated patches.npy this also give me the expected `n_clusters`

```
#patches=np.load(plot_dir+"/results/patches.npy")
i=0 
patch=patches[i]
pixel_values = patch.reshape((-1, 3))
pixel_values = np.float32(pixel_values)
n_clusters=10 
kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(pixel_values)
pred=kmeans.labels_
print(np.unique(pred))
```

<img width="1238" height="221" alt="Image" src="https://github.com/user-attachments/assets/79c56d18-f1f5-4fea-a4c6-64c197dbe53f" />

Continuing the pipeline also using your `patches.npy` as input, also only extracts mask 0, suggesting that the problem is not in the segmenting part, I think?

<img width="1244" height="508" alt="Image" src="https://github.com/user-attachments/assets/44be7842-f58b-4213-824b-5a60268d3bc6" />

**Comment by jinghuang-stats** [comment]
Hi Emma,
Thanks so much for testing! It's great to see that your `KMeans` function is running properly.
To pinpoint the issue, please use the `patches.npy` and `patches_info.csv` I've shared rather than your own generated patches -- it seems that the problem came from that step. Then, please rerun the `4.3 Segment each patch into masks` (the key idea is to replace your current `patches` and `patch_info` with the ones I've shared, and then rerun the following pipeline in the tutorial):
```
mph.step4_Segmentation(plot_dir=plot_dir, n_clusters=10, refine=True, refine_threshold=4)
mph.check_dic_list(plot_dir)
```
This step can take some time (around 2 hours). Once it finishes, you should find a newly generated `patch*_pred_refined.npy` in your `./seg_results` folder. You can quickly verify whether the segmentation produced 10 clusters by running:
```
patch_test=np.load("./seg_results/patch0_pred_refined.npy")
np.unique(patch_test)
```
If you see 10 unique clusters, everything is working correctly -- please continue with `4.4 Match masks across patches` and follow the remaining tutorial steps. Then you should be able to obtain multiple masks.
Please give this a try and see if it resolves the issue. It's always good to confirm this part independently first -- it'll help pinpoint where things go off track if the problem persists.

**Comment by Emmavb1988** [comment]
Thanks again for your quick replies. I think the problem remains in `4.3 Segment each patch into masks
`
here is my output after using your `patches.npy` and ` patches_info.csv` to run 

```
mph.step4_Segmentation(plot_dir=plot_dir, n_clusters=10, refine=True, refine_threshold=4)
mph.check_dic_list(plot_dir)
```

and then check the number of clusters using

```
patch_test=np.load("./seg_results/patch0_pred_refined.npy")
np.unique(patch_test)
```

<img width="1255" height="107" alt="Image" src="https://github.com/user-attachments/assets/ce1a1469-6990-4308-9f71-f135f641e644" />

Just to note, I did clear the seg_results folder before running 4.3 again, and the patchx_pred.npy (not refined) does have 10 unique clusters.

<img width="1232" height="123" alt="Image" src="https://github.com/user-attachments/assets/9b3e671a-f262-4b05-a26f-223b839deeb1" />

**Comment by jinghuang-stats** [comment]
Hi Emma,
Thanks so much for letting me know this important detail that `patchx_pred.npy` includes 10 clusters while `patchx_pred_refined.npy` includes only one cluster. This information was really helpful in pinpointing the issue within the `refine_labels` step of
`mph.step4_Segmentation(plot_dir=plot_dir, n_clusters=10, refine=True, refine_threshold=4)`
The issue arises from differences between older and newer versions of the **SciPy** package.

- In older versions (e.g., 1.7.1), the mode was retrieved using:
`scipy.stats.mode(clusters, axis=None).mode[0]`

- In newer versions (>=1.11.1), the mode was retrieved using:
`scipy.stats.mode(clusters, axis=None).mode`
Our current release was developed and tested under the newer SciPy. Therefore, running the pipeline in an environment with an older SciPy version (e.g., environment 3) can cause this issue in the refinement step.
To ensure consistent results, I've now removed environment 3 from the main page and updated the **package installation instructions** (`1.3 Anaconda`) in both `tutorial.ipynb` and `tutorial.md`. Please use Environment 1 or 2, both of which have been thoroughly tested and correctly produce 10 clusters in both `patchx_pred.npy` and `patchx_pred_refined.npy` (as shown in the figure).
<img width="787" height="101" alt="Image" src="https://github.com/user-attachments/assets/ad9151bb-4926-434e-8d87-be1f14ee0e5b" />
Hope this information helps resolve the issue on your end. Thanks again for your interest in MorphLink and valuable feedback.

**Comment by Emmavb1988** [comment]
Thank you very much, this indeed solved the issue and I was able to run the entire pipeline now!

I ran into two minor other things:

In preprocessing step **6.1** this gave an error
gene_adata.X = (np.array(gene_adata.X) if issparse(gene_adata.X) else np.array(gene_adata.X))
The matrices were alreay dense matrices so it did not matter, and adjusting the code to did not produce the error anymore:
`gene_adata.X = gene_adata.X.toarray() if issparse(gene_adata.X) else gene_adata.X`

I do get different louvain.clustering results but this might also be a package version issue at **6.2**
`img_adata=mph.louvain_clustering(input_adata=img_adata, pca_num=50, n_neighbors=5, resolution=0.5, pred_key="img_pred")`

Thank you again for your help and quick replies I'm now able to test the pipeline on my own data!

**Comment by jinghuang-stats** [comment]
Hi Emma,
I'm so glad to hear that the issue has been resolved and you're now able to run the pipeline successfully.
Thank you for your helpful feedback regarding the **6.1 Read in** `gene_adata` part, and I've updated the tutorial accordingly.
There might be some variations in clustering results due to the randomness in the PCA and clustering steps.
Thanks again for your interest in MorphLink, and I hope it can be helpful in your dataset.
